tempesta_fw/http_parser.c

```diff
@@ -3199,11 +3199,27 @@ tfw_http_parse_terminate(TfwHttpMsg *hm)
 	BUG_ON(!hm);
 	BUG_ON(!(TFW_CONN_TYPE(hm->conn) & Conn_Srv));
 
-	if (hm->parser.state == Resp_BodyUnlimRead) {
+	/*
+	 * Set Content-Length header to warn client about end of message.
+	 * Other option is to close connection to client. All the situations
+	 * when Content-Length header must not present in responce were
+	 * checked earlier. Refer to RFC 7230 3.3.3
+	 */
+	if (hm->parser.state == Resp_BodyUnlimRead
+	    || hm->parser.state == Resp_BodyUnlimStart)
+	{
+		char c_len[20] = {0};
+		int r;
+
 		BUG_ON(hm->body.flags & TFW_STR_COMPLETE);
 		hm->body.flags |= TFW_STR_COMPLETE;
 		hm->content_length = hm->body.len;
-		return true;
+		sprintf(c_len, "%lu", hm->content_length);
+		r = tfw_http_msg_hdr_xfrm(hm, "Content-Length:",
+					  sizeof("Content-Length:") - 1,
+					  c_len, strlen(c_len),
+					  TFW_HTTP_HDR_CONTENT_LENGTH, 0);
```

Comment:
Actually, this doesn't satisfy comment https://github.com/tempesta-tech/tempesta/issues/639#issuecomment-260643114 : according to the logic we must fully read while response before we can calculate Content-Length value. However, converting the whole response body to chunked encoding requirex XFRM logic for body which we don't have for now. Also we need to consider which chunks size we should use. So please mark the code with TODO comment mentioning issue #534 - we must rework the code.